crates/router-controller/src/controller.rs

```diff
@@ -125,10 +129,39 @@ impl Controller {
 
         match candidates.len() {
             0 => None,
-            1 => Some(&candidates[0].url),
+            1 => {
+                let worker = &candidates[0];
+                self.workers_rate.inc(&worker.url);
+                Some(&worker.url)
+            },
             len => {
-                let i: usize = rand::random();
-                Some(&candidates[i % len].url)
+                let key = (dataset.clone(), first_block);
+
+                let next_worker = if let Some(url) = self.request_tracker.get(&key) {
+                    let index = candidates.iter().position(|info| info.url == url);
+                    if let Some(index) = index {
+                        let worker = &candidates[(index + 1) % len];
+                        Some(worker)
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                };
+
+                let worker = if let Some(worker) = next_worker {
+                    worker
+                } else {
+                    let least_used_worker = candidates
+                        .iter()
+                        .min_by_key(|info| self.workers_rate.get_rate(&info.url));
```

Comment:
The current time is measured on every iteration. How about moving that `min` extraction inside `workers_rate` so that we only evaluate the current time once?